src/Paramore.Brighter.MongoDb.EntityFramework/MongoDbEntityFrameworkTransactionProvider.cs

```diff
@@ -0,0 +1,190 @@
+        /// </summary>
+        /// <returns>The NpgsqlConnection that is in use</returns>
+        public DbConnection GetConnection()
+        {
+            return _context.Database.GetDbConnection();
+        }
+    
+        /// <summary>
+        /// Get the current connection of the database context
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token</param>
+        /// <returns></returns>
+        public Task<DbConnection> GetConnectionAsync(CancellationToken cancellationToken = default)
+        {
+            var tcs = new TaskCompletionSource<DbConnection>();
+            tcs.SetResult(_context.Database.GetDbConnection());
+            return tcs.Task;
+        }
+    
+        /// <summary>
+        /// Get the ambient Transaction
+        /// </summary>
+        /// <returns>The NpgsqlTransaction</returns>
+        public IClientSessionHandle GetTransaction()
+        {
+            // If there is no current transaction, we create a new one
+            var currentTransaction = _context.Database.CurrentTransaction ?? _context.Database.BeginTransaction();
+            if (currentTransaction is not MongoTransaction mongoTransaction)
+            {
+                throw new InvalidOperationException("The current transaction is not a MongoTransaction");
+            }
+            // use reflection to access property named Session of type IClientSessionHandle that is in the mongoTransaction
+            // it is internal, so we need to use reflection to access it
+            // ideally we would use just DbTransaction directly, but it's currently impossible due to compatibility with the base mongodb outbox provider
+            if (MongoTransactionHelper.SessionProperty.GetValue(mongoTransaction) is not IClientSessionHandle session)
+            {
+                throw new InvalidOperationException("The current transaction does not have a session");
+            }
+            return session;
+        }
+    
+        public Task<IClientSessionHandle> GetTransactionAsync(CancellationToken cancellationToken = default)
+        {
+            var tcs = new TaskCompletionSource<IClientSessionHandle>();
+    
+            if (cancellationToken.IsCancellationRequested)
+            {
+                tcs.SetCanceled(CancellationToken.None);
+            }
+    
+            tcs.SetResult(GetTransaction());
+            return tcs.Task;
+        }
+    
+        [MemberNotNullWhen(true, nameof(_transaction))]
+        public bool HasOpenTransaction => _transaction is not null;
+    
+        public bool IsSharedConnection => true;
+    
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+    
+        protected void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+    
+            if (disposing)
+            {
+                _connection?.Dispose();
```

Comment:
@iancooper I'm not sure if we should dispose the connection and transaction in here, those should be managed by the DbContext itself i guess? This is a copy of postgres ef transaction provider